### PR TITLE
Systemic changes to support garter carriage on the full bed

### PR DESCRIPTION
**Pull request by clholgat**
https://github.com/AllYarnsAreBeautiful/ayab-firmware/issues/179

Things in this PR:
- Belt shift only set once per line (when the first magnet crosses the turn mark).
- Belt shift swapped (only for the first magnet of the garter carriage's that crosses the turn mark on the left)
- m_position increment/decrement no longer bounded, allowed to overflow correctly
- Offsets updated for G-carriage, L-carriage, K-carraige
- m_solenoid to set is now based on m_pixelToSet

Things that still need to be fixed (and will be before I request review):
- Verification that the 930 works
- 270 offsets and functionality need to be checked

After this PR, alignment issues will only ever show up at multiples of 8. I am so excited for this.

## Summary by CodeRabbit

- **New Features**
	- Introduced a new error signaling mechanism in the Beeper class, enhancing audible alerts for error conditions.
	- Added new constants for magnet sensor states to improve encoder functionality.
	- Implemented error handling capabilities in the GlobalBeeper class.

- **Improvements**
	- Simplified logic for position updates in the encoder, enhancing efficiency and robustness.
	- Adjusted parameters for encoder behavior related to machine configurations, improving operational accuracy.
	- Streamlined calculations for solenoid and pixel values in the knitting process, enhancing clarity and responsiveness.

- **Bug Fixes**
	- Improved handling of edge cases in encoder position updates, preventing erroneous state transitions.
	- Updated test cases to reflect changes in expected encoder behavior, ensuring accurate validation.

### CODE REVIEW

**Comment by jonathanperret** [comment]
First, congratulations on isolating this bug, it must not have been easy!

The following question is asked out of curiosity, feel free to disregard it.

> Also typecast increment/decrement so if it does overflow, it overflows correctly.
> 
> 1 is an int so m_position - 1 gets coerced to an int and then typecast back to a uint8_t upon assignment

I feel like I'm missing something here. I agree that `1` is treated as an `int`. And because `int` is wider than `uint8_t`, the operation `m_position - 1` is done by first widening the unsigned byte `m_position` to a signed 16-bit `int`, subtracting `1`, and casting the result back to an unsigned byte. But this still underflows "correctly", i.e. when `m_position` is `0`, the intermediate result is `(int)-1`, which casts to `(uint8_t)255` as expected.

Likewise when overflowing, if `m_position` is `255` the intermediate result is `256` which casts to `0`.

[Compiler explorer](https://godbolt.org/#z:OYLghAFBqd5TKALEBjA9gEwKYFFMCWALugE4A0BIEAZgQDbYB2AhgLbYgDkAjF%2BTXRMiAZVQtGIHgBYBQogFUAztgAKAD24AGfgCsp5eiyahUAUgBMAIUtXyKxqiIEh1ZpgDC6egFc2TA3cAGQImbAA5PwAjbFIpC3IAB3QlYhcmL19/A2TU5yEQsMi2GLieBIdsJ3SRIhZSIky/AJ57bEd8plr6okKI6Nj4%2BzqGpuzWpRHe0P6SwfKASnt0H1JUTi4iAE9E7BwaAGofJlTgMMwD1CR6o9CiAA4AfSIzAGYbLQBBMy%2BfO6eiAdQqgIH9hACDkw/AsDmYAOwfT4HZGQvyw14AEVRbFh1gOPDeiJRB1I2CIqyY2MJP2%2BcIxNJpYIezwOOBBTIhULYMPhRJRXPRWIFAFp8dSvsTSeTSJSueLafSvlwlvRuABWfgBXj8dDcDy2WwHJQrNbYXGvPjkIjaZVLADWIFerwAdE63e6PQA2QzcaT8NggNVachanTkXVcfhKEDB61cHRLOCwFAYHD4YhkSjUOiMVgcbiWwTCMQSTgyOTCZRqTTxvQgBJGExoA3WNoddJuJiebzNQJdvrFUpDXJpIRjFpJFKjpgDgZlCrtKqdbqjHvjNtLmpTWdzefDHrjgyTHo7oeLZardbcba7fZHE4EM57S7XUi3cHPeWM/4s4Ggn%2BAlyPIIjSxICm8Qpooa/4fkQCwEu8oEolKFJUohXzwoq3y/ABrJVDBzKAdCsIgRK/JQZi2IHKKBEAvB8qSmSqFyuhCo0sqPpcBqIY2uGeotlYRomus5oWPwcYJksSDYCwOBxBAKq%2Bv6gbBqGOrcFGMZWjaCz2o6LoeoZbreqqXCvJqvERuJOnkEmCDwBAKboGwiQMLEWYQBgLluXELAAG6kMAMg8MGOZELE0YQFEvFRKE9RbAW/CxawpBbAA8lEuhLol5BeRwwhpUw9AJbW5A4FEPjAB4Ej0NG2pldgbDGMAkilYQpLVH52B1WG2DqFUPjhTldztLx9AEFEpDxV4OC8UQpAEAG9VdaQUQpNgGKNc141NrWSw0EYwBKAAagQ2AAO5pbsWqFvIJaSCFFaKCoGi8forSNqYAmGBN0aQEs6CJJ0dU6itC04H9CkbtUrgQO4h6tMEMyDvMk55OkCNo9Op6o5UMNdFMmN48u27I3OR6E2uE7Hg0ONlEsxqXmWilcRZpURgcnzHQASgcwCoKg%2BLSM6PDOloBwQOmJBvpYFoLNZe16U6rpGYZJlKeQAbSPcIsWOU%2Bt64b5Q8ezGn2FpEkcXZjloM5rmMBQVCeXbPkgP5gXBaFDDhaQkXRaVyXxTlgepRlWVODleXMEQhXFbx5WVdV9C1TlOBNSYrVhu1m5dT1/B9QNQ31SNplhuNk3TVgGxhvNi05Sta0qJt6ctaEoB7QIh0nWdl3XTlRaiOID2hfIVavaV%2BivIYzXNtYtg/VEkMA0D6Qg%2BGYMEBD8AM4u%2BOdt2WQBBYshI0U5OPSOnSHsfWOdHTUjBsTW4HlT9ayE/QgrtMZ%2B7g/%2B6rofN%2B/9v6zDPFoBmwlOAWBZtxNSfEuCcx5nzAWQsRZiwllLMgol5baUVuQaSslBhQ1Mn6TWgZRb3DhJQ6hVDaH3HuCbMMVlzaxh0krAyqsvScXMow9SkZcEJlsogG2hAaA0A8gPe6ZYR6VhejWMuSBowfUUaImgN5ODBnoIovq80WCA06GbUgijHTkCMUoVR6itJmJ0VNfRHZIwwLZkw7gGICBiIOKdC6sREG835oLGQaDxZ%2BSUD45B/jhaixrmw8gDplacI9JxUhcDmHRlYXg0yYleHwIVpJcgK1UiuGkEAA%3D%3D%3D) seems to show that the same assembly code is generated with or without the `(uint8_t)` cast.

Did you observe different behavior when adding the `(uint8_t)` cast? 

Finally, as suggested by @X-sam on Discord, wouldn't it be safer all around to upgrade `m_position` to an `int`?

**Comment by jonathanperret** [comment]
> Finally, as suggested by @X-sam on Discord, wouldn't it be safer all around to upgrade `m_position` to an `int`?

Just thought of a potential downside to making `m_position` an `int`: updates wouldn’t be atomic anymore, so accesses outside of an ISR would require more care.

File: src/ayab/encoders.cpp [hunk 1]
```diff
@@ -133,8 +133,8 @@ void Encoders::encA_rising() {
   m_direction = digitalRead(ENC_PIN_B) != 0 ? Direction_t::Right : Direction_t::Left;
 
   // Update carriage position
-  if ((Direction_t::Right == m_direction) && (m_position < END_RIGHT[static_cast<uint8_t>(m_machineType)])) {
-    m_position = m_position + 1;
+  if ((Direction_t::Right == m_direction)) {
```
Comment (jonathanperret) [inline src/ayab/encoders.cpp:136]: Yes, removing this test on machine limits is a good call. Blindly ignoring that the machine did move out of the expected bounds wasn’t going  to be help 😬 
In the future, maybe reporting the unexpected situation somehow (going into an error state?) could help catch unwarranted assumptions earlier.

**Comment by clholgat** [comment]
Not ready for review, still sorting out RTL offsets and then need to do a lot of cleanup.

File: src/ayab/encoders.cpp [hunk 2]
```diff
@@ -155,6 +150,38 @@ void Encoders::encA_rising() {
...
+            m_beltShift = BeltShift::Regular;
+          }
+        }
+      }
+    }
+
+    // The garter carriage has a second set of magnets that are going to
+    // pass the sensor and will reset state incorrectly if allowed to
+    // continue.
+    if (m_carriage == Carriage_t::Garter) {
```
Comment (X-sam) [inline src/ayab/encoders.cpp:176]: Nit- in some of your comparisons you put the constant first and in some you use the variable first. I think we should prefer constant first but either way, surely we want to be consistent?

Comment (jonathanperret) [reply src/ayab/encoders.cpp:176]: +1 for consistency.

Although if we have to settle on one style, I'd rather have the constant on the right as above. The safety advantages of the reversed order are inexistent, since if `if (m_carriage = Carriage::Garter)` is written, the compiler will both raise a warning (which is treated as an error) and a type error because enum classes can't be implicitly converted to booleans.

As a side note, it seems to me the `Carriage_t` typedef (and others like it) only still exists for compatibility with existing code. `Carriage::Garter` works just as fine and seems cleaner; most code still uses the `_t` variants but there are many occurrences of `BeltShift::` for example — here too, consistency would be nice.

File: src/ayab/encoders.cpp [hunk 3]
```diff
@@ -176,16 +203,25 @@ void Encoders::encA_rising() {
     } else if (m_carriage != detected_carriage && m_position > start_position) {
       m_carriage = Carriage_t::Garter;
 
+      // We swap the belt shift for the g-carriage because the point of work for 
+      // the g-carraige is 13 needles behind the first magnet which puts it in a different
+      // belt shift.
+      // And we need to know the belt shift when the point of work is at needle 0.
+      // Conveniently, the magnet distance on the K and L carraiges puts the point of
```
Comment (jonathanperret) [inline src/ayab/encoders.cpp:210]: **carraige** again
```suggestion
      // Conveniently, the magnet distance on the K and L carriages puts the point of
```

File: src/ayab/knitter.cpp [hunk 4]
```diff
@@ -388,60 +389,39 @@ void Knitter::reqLine(uint8_t lineNumber) {
...
+
+  bool beltShift = BeltShift_t::Shifted == m_beltShift;
 
   switch (m_direction) {
   // calculate the solenoid and pixel to be set
   // implemented according to machine manual
   // magic numbers from machine manual
   case Direction_t::Right:
     startOffset = getStartOffset(Direction_t::Left);
+    m_pixelToSet = m_position - startOffset;
```
Comment (jonathanperret) [inline src/ayab/knitter.cpp:401]: It looks like this line is identical in both branches of the `switch`, maybe it could be moved below it?

One other potential simplification would be to switch the direction-indexed values in `START_OFFSET` (unless there's a good reason for having `START_OFFSET[Right]` designate the offset for moving to the _left_, but I can't see it) so that you could simply write `startOffset = getStartOffset(m_direction)` and move that out of the `switch` as well.

**Comment by dl1com** [comment]
Tests still failing after rebase

File: src/ayab/encoders.cpp [hunk 5]
```diff
@@ -155,6 +150,38 @@ void Encoders::encA_rising() {
...
+    // Only set the belt shift the first time a magnet passes the turn mark.
+    // Headed to the right.
+    if (!m_passedLeft && Direction_t::Right == m_direction) {
+      // Belt shift signal only decided in front of hall sensor
+      m_beltShift = digitalRead(ENC_PIN_C) != 0 ? BeltShift::Shifted : BeltShift::Regular;
+      m_passedLeft = true;
+
+      if (Carriage_t::Garter == m_carriage) {
+        // This has to be the first magnet and the belt shift needs to be swapped
+        // But only for the G-carraige
```
Comment (jonathanperret) [inline src/ayab/encoders.cpp:162]: oops, *carraige*, looks like it's the last one.
```suggestion
        // But only for the G-carriage
```

File: src/ayab/knitter.cpp (hunk 4 above)
Comment (clholgat) [reply src/ayab/knitter.cpp:399]: I'll move the m_pixelToSet out but I'm going to leave the start offset stuff for now.

**Comment by clholgat** [comment]
Can I get a review?

File: src/ayab/encoders.cpp [hunk 6]
```diff
@@ -155,6 +150,38 @@ void Encoders::encA_rising() {
       (hallValue > FILTER_L_MAX[static_cast<uint8_t>(m_machineType)])) {
     m_hallActive = Direction_t::Left;
 
+    // Only set the belt shift the first time a magnet passes the turn mark.
+    // Headed to the right.
+    if (!m_passedLeft && Direction_t::Right == m_direction) {
+      // Belt shift signal only decided in front of hall sensor
+      m_beltShift = digitalRead(ENC_PIN_C) != 0 ? BeltShift::Shifted : BeltShift::Regular;
```
Comment (X-sam) [inline src/ayab/encoders.cpp:157]: Personally I find this line a bit dense, I'd refactor to something like, `if (digital read()) m_beltshift = shifted`.

File: src/ayab/encoders.cpp [hunk 7]
```diff
@@ -203,8 +239,13 @@ void Encoders::encA_falling() {
   m_direction = digitalRead(ENC_PIN_B) ? Direction_t::Left : Direction_t::Right;
 
   // Update carriage position
-  if ((Direction_t::Left == m_direction) && (m_position > END_LEFT[static_cast<uint8_t>(m_machineType)])) {
-    m_position = m_position - 1;
+  if (Direction_t::Left == m_direction) {
+    m_position = m_position - (uint8_t) 1;
```
Comment (X-sam) [inline src/ayab/encoders.cpp:243]: Do we really need to cast the 1? Oof.

File: src/ayab/encoders.cpp [hunk 8]
```diff
@@ -219,15 +260,26 @@ void Encoders::encA_falling() {
   if (hallValueSmall || hallValue > FILTER_R_MAX[static_cast<uint8_t>(m_machineType)]) {
     m_hallActive = Direction_t::Right;
 
-    // The garter carriage has a second set of magnets that are going to
+    // Only set the belt shift when the first magnet passes the turn mark.
+    // Headed to the left.
+    if (!m_passedRight && Direction_t::Left == m_direction) {
+      // Belt shift signal only decided in front of hall sensor
+      m_beltShift = digitalRead(ENC_PIN_C) != 0 ? BeltShift::Regular : BeltShift::Shifted;
```
Comment (X-sam) [inline src/ayab/encoders.cpp:267]: Same comment as the left.

**Review by X-sam — APPROVED** [review]
Lgtm

File: src/ayab/encoders.cpp (hunk 6 above)
Comment (clholgat) [reply src/ayab/encoders.cpp:157]: We'd still need to set the BeltShift::Regular, the default is BeltShift::Unknown.

Honestly, the ternary operator is the best option here.

File: src/ayab/encoders.cpp (hunk 8 above)
Comment (clholgat) [reply src/ayab/encoders.cpp:267]: Same response

File: src/ayab/encoders.cpp (hunk 7 above)
Comment (clholgat) [reply src/ayab/encoders.cpp:243]: Let me double check

File: src/ayab/encoders.cpp (hunk 6 above)
Comment (X-sam) [reply src/ayab/encoders.cpp:157]: I don't care enough to fight this one, happy to let it go :)
